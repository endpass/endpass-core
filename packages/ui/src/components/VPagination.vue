<template>
  <div class="v-pagination">
    <div class="v-pagination-navigation">
      <span class="v-pagination-node">Page {{ currentPage }}</span>
      <a
        :disabled="!hasBack"
        class="v-pagination-node"
        data-test="pagination-prev"
        @click.prevent="goToPrev()"
      >
        <span class="v-pagination-triangle-left"/>
      </a>
      <a
        v-if="showFirstPage"
        class="v-pagination-node"
        data-test="btn-goto-first"
        @click.prevent="goToFirst()"
      >
        1
      </a>
      <a
        v-if="isShowFirstPageDots"
        class="v-pagination-node"
        data-test="dots-left"
        @click.prevent="goToFirst()"
      >...</a>

      <a
        class="v-pagination-node"
        v-for="num in allVisiblePages"
        data-test="btn-goto-page"
        @click.prevent="goToPage(num)"
      >
        {{ num }}
        <span 
          :class="{ 'v-pagination-active-foot': currentPage === num }" 
        />
      </a>

      <a
        v-if="showLastPageDots"
        class="v-pagination-node"
        data-test="dots-right"
        @click.prevent="goToLast()"
      >...</a>
      <a
        data-test="btn-goto-last"
        v-if="showLastPage"
        class="v-pagination-node"
        @click.prevent="goToLast()"
      >
        {{ totalPages }}
      </a>
      <a
        :disabled="!hasForward"
        class="v-pagination-node"
        data-test="pagination-next"
        @click.prevent="goToNext()"
      >
        <span class="v-pagination-triangle-right"/>
      </a>
    </div>
  </div>
</template>

<script>
export default {
  name: 'VPagination',
  props: {
    offset: {
      type: Number,
      required: true,
    },
    limit: {
      required: true,
      validator: (value) => {
        return !!value && Number(value) > 0; // 1+
      }
    },
    total: {
      type: Number,
      default: 0,
    },
    pagesVisible: {
      default: 0,
      validator: (value) => {
        return value === 0 || ((value % 2) === 1); // 0, 1, 3, 5 ...
      },
    },
  },
  computed: {
    currentPage() {
      return Math.floor(this.offset / this.limit) + 1;
    },
    halfPagesVisible() {
      return Math.floor(this.pagesVisible / 2);
    },
    totalPages() {
      return Math.ceil(this.total / this.limit);
    },

    allVisiblePages() {
      // visible = 3
      // [[1], 2, 3] = []
      // [1, [2], 3] = [1]
      // [6, 7, [8]] = [6, 7]

      const startPointBefore = Math.max(1, this.currentPage - this.halfPagesVisible);

      const maxPages = this.totalPages
        ? this.totalPages + 1
        : this.currentPage + this.pagesVisible;

      const startPointByVisiblePages = Math.max(1, maxPages - this.pagesVisible);
      const startPoint = Math.min(startPointBefore, startPointByVisiblePages);
      const endPoint = Math.min(maxPages, startPoint + this.pagesVisible);

      const countTotalPages = endPoint - startPoint;

      return Array.from(Array(countTotalPages), (d, i) => {
        return startPoint + i;
      });
    },

    isInfinityPages() {
      return this.total === 0;
    },
    hasVisiblePages() {
      return !!this.pagesVisible;
    },
    hasBack() {
      return this.currentPage !== 1;
    },
    hasForward() {
      if (this.isInfinityPages) {
        return true; // infinity next
      }

      return this.totalPages > this.currentPage;
    },
    showFirstPage() {
      if (!this.hasVisiblePages) {
        return false;
      }

      return !this.allVisiblePages.includes(1);
    },
    isShowFirstPageDots(){
      if (!this.hasVisiblePages) {
        return false;
      }

      if (this.allVisiblePages.includes(2)) {
        return false;
      }

      return this.currentPage > 2;
    },
    showLastPage() {
      if (!this.hasVisiblePages) {
        return false;
      }

      return !this.isInfinityPages && !this.allVisiblePages.includes(this.totalPages);
    },
    showLastPageDots() {
      if (!this.hasVisiblePages) {
        return false;
      }

      const nearPageNumber = this.totalPages - 1;
      if (this.allVisiblePages.includes(nearPageNumber)) {
        return false;
      }

      if (this.isInfinityPages) {
        return true;
      }

      return this.totalPages > 2;
    }
  },
  methods: {
    goToFirst() {
      this.goToPage(1);
    },

    goToLast() {
      this.goToPage(this.totalPages);
    },

    goToPrev() {
      this.goToPage(this.currentPage - 1);
    },

    goToNext() {
      this.goToPage(this.currentPage + 1);
    },

    goToPage(pageNumber) {
      if (pageNumber > 0 && this.currentPage === pageNumber) {
        return;
      }

      const nextOffset = (pageNumber - 1) * this.limit;

      this.$emit('offset', nextOffset);
    },
  },
};
</script>

<style lang="scss">
  .v-pagination {
    display: flex;
  }

  .v-pagination-navigation {
    display: flex;
    flex-flow: row wrap;
    justify-content: flex-start;
  }

  .v-pagination-node {
    position: relative;
    height: 48px;
    width: 48px;
    white-space: nowrap;
    align-items: center;
    cursor: pointer;
    display: flex;
    justify-content: center;
    flex-flow: column nowrap;
    padding: 12px;

    &[disabled="disabled"] {
      pointer-events: none;
      cursor: default;
      opacity: 0.3;
    }
  }

  .v-pagination-triangle-left {
    width: 0;
    height: 0;
    border-style: solid;
    border-width: 5px 7px 5px 0;
    border-color: transparent #000000 transparent transparent;
  }

  .v-pagination-triangle-right {
    width: 0;
    height: 0;
    border-style: solid;
    border-width: 5px 0 5px 7px;
    border-color: transparent transparent transparent #000000;
  }

  .v-pagination-active-foot {
    position: absolute;
    bottom: 0;
    left: 50%;
    margin-left: -8px;
    height: 4px;
    background-color: #6E32C9;
    width: 16px;
  }
</style>
