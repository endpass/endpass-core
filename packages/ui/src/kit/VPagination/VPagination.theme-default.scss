.v-pagination.theme-default {
  display: flex;
}

.v-pagination.theme-default .v-pagination-navigation {
  display: flex;
  flex-flow: row wrap;
  justify-content: flex-start;
}

.v-pagination.theme-default .v-pagination-node {
  position: relative;
  height: 48px;
  width: 48px;
  white-space: nowrap;
  align-items: center;
  display: flex;
  justify-content: center;
  flex-flow: column nowrap;
  padding: 12px;

  &[disabled="disabled"] {
    pointer-events: none;
    cursor: default;
    opacity: 0.3;
  }
}

a.v-pagination-node{
  cursor: pointer;
}

.v-pagination.theme-default .v-pagination-triangle-left {
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 5px 7px 5px 0;
  border-color: transparent #000000 transparent transparent;
}

.v-pagination.theme-default .v-pagination-triangle-right {
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 5px 0 5px 7px;
  border-color: transparent transparent transparent #000000;
}

.v-pagination.theme-default .v-pagination-active-foot {
  position: absolute;
  bottom: 0;
  left: 50%;
  margin-left: -8px;
  height: 4px;
  background-color: $primary-7;
  width: 16px;
}
