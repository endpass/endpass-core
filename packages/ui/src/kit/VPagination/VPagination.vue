<template>
  <div
    class="v-pagination"
    :class="themeCssClass"
  >
    <div class="v-pagination-navigation">
      <span
        v-if="showCurrentPageTitle"
        class="v-pagination-node"
      >{{ title }} {{ currentPage }}</span>
      <a
        :disabled="!hasBack"
        class="v-pagination-node"
        data-test="pagination-prev"
        @click.prevent="goToPrev()"
      >
        <span class="v-pagination-triangle-left" />
      </a>
      <a
        v-if="showFirstPage"
        class="v-pagination-node"
        data-test="btn-goto-first"
        @click.prevent="goToFirst()"
      >
        1
      </a>
      <span
        v-if="isShowFirstPageDots"
        class="v-pagination-node"
        data-test="dots-left"
      >...</span>

      <a
        v-for="num in allVisiblePages"
        :key="num"
        class="v-pagination-node"
        data-test="btn-goto-page"
        @click.prevent="goToPage(num)"
      >
        {{ num }}
        <span :class="{ 'v-pagination-active-foot': currentPage === num }" />
      </a>

      <span
        v-if="showLastPageDots"
        class="v-pagination-node"
        data-test="dots-right"
      >...</span>
      <a
        v-if="showLastPage"
        data-test="btn-goto-last"
        class="v-pagination-node"
        @click.prevent="goToLast()"
      >
        {{ totalPages }}
      </a>
      <a
        :disabled="!hasForward"
        class="v-pagination-node"
        data-test="pagination-next"
        @click.prevent="goToNext()"
      >
        <span class="v-pagination-triangle-right" />
      </a>
    </div>
  </div>
</template>

<script>
import ThemeMixin from '@/mixins/ThemeMixin';

export default {
  name: 'VPagination',
  props: {
    title: {
      type: String,
      default: 'Page',
    },
    offset: {
      type: Number,
      required: true,
    },
    limit: {
      required: true,
      validator: value => !!value && Number(value) > 0, // 1+
    },
    total: {
      type: Number,
      default: 0,
    },
    pagesVisible: {
      default: 0,
      validator: value => value === 0 || value % 2 === 1, // 0, 1, 3, 5 ...
    },
  },
  computed: {
    currentPage() {
      return Math.floor(this.offset / this.limit) + 1;
    },
    halfPagesVisible() {
      return Math.floor(this.pagesVisible / 2);
    },
    totalPages() {
      return Math.ceil(this.total / this.limit);
    },

    allVisiblePages() {
      // visible = 3
      // [[1], 2, 3] = []
      // [1, [2], 3] = [1]
      // [6, 7, [8]] = [6, 7]

      const startPointBefore = Math.max(
        1,
        this.currentPage - this.halfPagesVisible,
      );

      const maxPages = this.totalPages
        ? this.totalPages + 1
        : this.currentPage + this.pagesVisible;

      const startPointByVisiblePages = Math.max(
        1,
        maxPages - this.pagesVisible,
      );
      const startPoint = Math.min(startPointBefore, startPointByVisiblePages);
      const endPoint = Math.min(maxPages, startPoint + this.pagesVisible);

      const countTotalPages = endPoint - startPoint;

      return Array.from(Array(countTotalPages), (d, i) => startPoint + i);
    },

    isInfinityPages() {
      return this.total === 0;
    },
    hasVisiblePages() {
      return !!this.pagesVisible;
    },
    hasBack() {
      return this.currentPage !== 1;
    },
    hasForward() {
      if (this.isInfinityPages) {
        return true; // infinity next
      }

      return this.totalPages > this.currentPage;
    },
    showFirstPage() {
      if (!this.hasVisiblePages) {
        return false;
      }

      return !this.allVisiblePages.includes(1);
    },
    isShowFirstPageDots() {
      if (!this.hasVisiblePages) {
        return false;
      }

      if (this.allVisiblePages.includes(2)) {
        return false;
      }

      return this.currentPage > 2;
    },
    showLastPage() {
      if (!this.hasVisiblePages) {
        return false;
      }

      return (
        !this.isInfinityPages && !this.allVisiblePages.includes(this.totalPages)
      );
    },
    showLastPageDots() {
      if (!this.hasVisiblePages) {
        return false;
      }

      const nearPageNumber = this.totalPages - 1;
      if (this.allVisiblePages.includes(nearPageNumber)) {
        return false;
      }

      if (this.isInfinityPages) {
        return true;
      }

      return this.totalPages > 2;
    },
    showCurrentPageTitle() {
      return this.pagesVisible === 0;
    },
  },
  methods: {
    goToFirst() {
      this.goToPage(1);
    },

    goToLast() {
      this.goToPage(this.totalPages);
    },

    goToPrev() {
      this.goToPage(this.currentPage - 1);
    },

    goToNext() {
      this.goToPage(this.currentPage + 1);
    },

    goToPage(pageNumber) {
      if (pageNumber > 0 && this.currentPage === pageNumber) {
        return;
      }

      const nextOffset = (pageNumber - 1) * this.limit;

      this.$emit('offset', nextOffset);
    },
  },
  mixins: [ThemeMixin],
};
</script>
