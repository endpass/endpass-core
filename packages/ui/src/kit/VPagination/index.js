export { default } from './VPagination.vue';
