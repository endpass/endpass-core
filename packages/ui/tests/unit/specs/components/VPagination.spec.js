import { shallowMount } from '@vue/test-utils';

import VPagination from '@/components/VPagination';

describe('VPagination', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(VPagination);
    jest.clearAllMocks();
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('VPagination');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });
  });

  describe('behaviour', () => {
    it('should render only "next"', () => {
      expect(wrapper.find('[data-test=pagination-next]').exists()).toBeTruthy();
      expect(wrapper.find('[data-test=pagination-prev]').exists()).toBeFalsy();
    });

    it('should render with "back"', () => {
      wrapper.setProps({
        offset: 10,
      });

      expect(wrapper.find('[data-test=pagination-next]').exists()).toBeTruthy();
      expect(wrapper.find('[data-test=pagination-prev]').exists()).toBeTruthy();
    });

    it('should render without "next"', () => {
      wrapper.setProps({
        offset: 10,
        limit: 10,
        total: 12,
      });

      expect(wrapper.find('[data-test=pagination-next]').exists()).toBeFalsy();
      expect(wrapper.find('[data-test=pagination-prev]').exists()).toBeTruthy();
    });

    it('should emit correct next offset', () => {
      wrapper.setProps({
        offset: 0,
        limit: 10,
      });

      wrapper.find('[data-test=pagination-next]').trigger('click');

      expect(wrapper.emitted().offset).toEqual([[10]]);
    });

    it('should emit correct prev offset', () => {
      wrapper.setProps({
        offset: 10,
        limit: 10,
        total: 12,
      });

      wrapper.find('[data-test=pagination-prev]').trigger('click');

      expect(wrapper.emitted().offset).toEqual([[0]]);
    });
  });
});
