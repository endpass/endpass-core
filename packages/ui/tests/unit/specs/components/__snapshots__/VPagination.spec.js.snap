// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VPagination render should be with default state 1`] = `
<div class="v-pagination">
  <div class="v-pagination-navigation"><span class="v-pagination-node">Page 1</span> <a disabled="disabled" data-test="pagination-prev" class="v-pagination-node"><span class="v-pagination-triangle-left"></span></a>
    <!---->
    <!---->
    <!---->
    <!----> <a data-test="pagination-next" class="v-pagination-node"><span class="v-pagination-triangle-right"></span></a></div>
</div>
`;

exports[`VPagination render should be with defined total 1`] = `
<div class="v-pagination">
  <div class="v-pagination-navigation"><span class="v-pagination-node">Page 6</span> <a data-test="pagination-prev" class="v-pagination-node"><span class="v-pagination-triangle-left"></span></a> <a data-test="btn-goto-first" class="v-pagination-node">
      1
    </a> <a data-test="dots-left" class="v-pagination-node">...</a> <a data-test="btn-goto-page" class="v-pagination-node">
      5
      <span class=""></span></a><a data-test="btn-goto-page" class="v-pagination-node">
      6
      <span class="v-pagination-active-foot"></span></a><a data-test="btn-goto-page" class="v-pagination-node">
      7
      <span class=""></span></a> <a data-test="dots-right" class="v-pagination-node">...</a> <a data-test="btn-goto-last" class="v-pagination-node">
      10
    </a> <a data-test="pagination-next" class="v-pagination-node"><span class="v-pagination-triangle-right"></span></a></div>
</div>
`;

exports[`VPagination render should be with defined total in end 1`] = `
<div class="v-pagination">
  <div class="v-pagination-navigation"><span class="v-pagination-node">Page 11</span> <a data-test="pagination-prev" class="v-pagination-node"><span class="v-pagination-triangle-left"></span></a> <a data-test="btn-goto-first" class="v-pagination-node">
      1
    </a> <a data-test="dots-left" class="v-pagination-node">...</a> <a data-test="btn-goto-page" class="v-pagination-node">
      8
      <span class=""></span></a><a data-test="btn-goto-page" class="v-pagination-node">
      9
      <span class=""></span></a><a data-test="btn-goto-page" class="v-pagination-node">
      10
      <span class=""></span></a>
    <!---->
    <!----> <a data-test="pagination-next" class="v-pagination-node" disabled="disabled"><span class="v-pagination-triangle-right"></span></a></div>
</div>
`;

exports[`VPagination render should be with pagesVisible 1`] = `
<div class="v-pagination">
  <div class="v-pagination-navigation"><span class="v-pagination-node">Page 5</span> <a data-test="pagination-prev" class="v-pagination-node"><span class="v-pagination-triangle-left"></span></a> <a data-test="btn-goto-first" class="v-pagination-node">
      1
    </a> <a data-test="dots-left" class="v-pagination-node">...</a> <a data-test="btn-goto-page" class="v-pagination-node">
      4
      <span class=""></span></a><a data-test="btn-goto-page" class="v-pagination-node">
      5
      <span class="v-pagination-active-foot"></span></a><a data-test="btn-goto-page" class="v-pagination-node">
      6
      <span class=""></span></a> <a data-test="dots-right" class="v-pagination-node">...</a>
    <!----> <a data-test="pagination-next" class="v-pagination-node"><span class="v-pagination-triangle-right"></span></a></div>
</div>
`;
